// @tableland/sdk/dist/cjs/lib/eth-calls.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.lockController = exports.getController = exports.setController = exports.runSql = exports.createTable = void 0;
const evm_1 = require("@tableland/evm");
const util_js_1 = require("./util.js");
function createTable(query) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        this.signer = (_a = this.signer) !== null && _a !== void 0 ? _a : (yield (0, util_js_1.getSigner)());
        const address = yield this.signer.getAddress();
        const contractAddress = this.options.contract;
        const contract = evm_1.TablelandTables__factory.connect(contractAddress, this.signer);
        const opts = yield getOverrides(this.signer);
        const tx = yield contract.createTable(address, query, opts);
        return yield tx.wait();
    });
}
exports.createTable = createTable;
function runSql(tableId, query) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        this.signer = (_a = this.signer) !== null && _a !== void 0 ? _a : (yield (0, util_js_1.getSigner)());
        const address = yield this.signer.getAddress();
        const contractAddress = this.options.contract;
        const contract = evm_1.TablelandTables__factory.connect(contractAddress, this.signer);
        const opts = yield getOverrides(this.signer);
        const tx = yield contract.runSQL(address, tableId, query, opts);
        return yield tx.wait();
    });
}
exports.runSql = runSql;
function setController(tableId, controller) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        this.signer = (_a = this.signer) !== null && _a !== void 0 ? _a : (yield (0, util_js_1.getSigner)());
        const caller = yield this.signer.getAddress();
        const contractAddress = this.options.contract;
        const contract = evm_1.TablelandTables__factory.connect(contractAddress, this.signer);
        const opts = yield getOverrides(this.signer);
        const tx = yield contract.setController(caller, tableId, controller, opts);
        return yield tx.wait();
    });
}
exports.setController = setController;
function getController(tableId) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        this.signer = (_a = this.signer) !== null && _a !== void 0 ? _a : (yield (0, util_js_1.getSigner)());
        const contractAddress = this.options.contract;
        const contract = evm_1.TablelandTables__factory.connect(contractAddress, this.signer);
        return yield contract.getController(tableId);
    });
}
exports.getController = getController;
function lockController(tableId) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        this.signer = (_a = this.signer) !== null && _a !== void 0 ? _a : (yield (0, util_js_1.getSigner)());
        const caller = yield this.signer.getAddress();
        const contractAddress = this.options.contract;
        const contract = evm_1.TablelandTables__factory.connect(contractAddress, this.signer);
        const opts = yield getOverrides(this.signer);
        const tx = yield contract.lockController(caller, tableId, opts);
        return yield tx.wait();
    });
}
exports.lockController = lockController;
function getOverrides(signer) {
    var _a;
    return __awaiter(this, void 0, void 0, function* () {
        // Hack: Revert to gasPrice to avoid always underpriced eip-1559 transactions on Polygon
        const opts = {};
        const network = yield ((_a = signer.provider) === null || _a === void 0 ? void 0 : _a.getNetwork());
        if ((network === null || network === void 0 ? void 0 : network.chainId) === 137) {
            const feeData = yield signer.getFeeData();
            if (feeData.gasPrice) {
                opts.gasPrice =
                    Math.floor(feeData.gasPrice.toNumber() * 1.1) || undefined;
            }
        }
        return opts;
    });
}
